### PR TITLE
Changing mempool_getPendingTransactions to include future transactions.

### PR DESCRIPTION
**Pull request by MSNTCS**
The new change is made. The change is supposed to include the future queue in pending transactions as well. Previously, we only had the current queue. Now, `mempool_getPendingTransactions` can return either the `current queue` or both the `current` and `future queues`. In order to have `future queue`, the boolean `future_included` parameter is considered.

### CODE REVIEW

File: test/src/helper/spawn.ts [hunk 1]
```diff
@@ -859,7 +859,7 @@ export default class CodeChain {
     public async waitForTermChange(target: number, timeout?: number) {
         const start = Date.now();
         while (true) {
-            const termMetadata = await stake.getTermMetadata(this.sdk);
+            const termMetadata = await stake.getTermMetadata(this.sdk as any);
```
Comment (joojis) [inline test/src/helper/spawn.ts]: It seems to be unrelated. Please remove this change from the PR.

File: test/src/e2e/mempool.test.ts [hunk 2]
```diff
@@ -46,26 +47,54 @@ describe("Sealing test", function() {
...
-        expect(await node.sdk.rpc.chain.getSeq(faucetAddress)).to.equal(seq);
-        await node.sendPayTx({ seq: seq + 2 });
-        expect(await node.sdk.rpc.chain.getSeq(faucetAddress)).to.equal(seq);
-        await node.sendPayTx({ seq: seq + 1 });
-        expect(await node.sdk.rpc.chain.getSeq(faucetAddress)).to.equal(seq);
-        await node.sendPayTx({ seq: seq });
-        expect(await node.sdk.rpc.chain.getSeq(faucetAddress)).to.equal(
-            seq + 4
-        );
+    it.only("all transaction in both queues should be included", async function() {
```
Comment (joojis) [inline test/src/e2e/mempool.test.ts]: You need to remove `only` here.

Comment (joojis) [reply test/src/e2e/mempool.test.ts]: Plus, in this case you need to add a new test,  not changing the existing one.

File: test/src/helper/spawn.ts (hunk 1 above)
Comment (MSNTCS) [reply test/src/helper/spawn.ts]: Done ✅

File: test/src/e2e/mempool.test.ts (hunk 2 above)
Comment (MSNTCS) [reply test/src/e2e/mempool.test.ts]: Done ✅

File: core/src/transaction.rs [hunk 3]
```diff
@@ -240,6 +240,7 @@ pub struct LocalizedTransaction {
 impl LocalizedTransaction {
     /// Returns transaction signer.
     /// Panics if `LocalizedTransaction` is constructed using invalid `UnverifiedTransaction`.
+
```
Comment (joojis) [inline core/src/transaction.rs]: Could you remove this change?

This is the only change in the file but unnecessary.

File: core/src/client/test_client.rs [hunk 4]
```diff
@@ -542,10 +542,17 @@ impl BlockChainClient for TestBlockChainClient {
         self.miner.ready_transactions(range)
     }
 
+    fn future_ready_transactions(&self, range: Range<u64>) -> PendingSignedTransactions {
+        self.miner.future_ready_transactions(range)
```
Comment (joojis) [inline core/src/client/test_client.rs]: I'm curious if there is a better name.

I think, in this context, "ready" means the "current". For me, it's hard to make guess the behavior from this name.

File: core/src/miner/mem_pool.rs [hunk 5]
```diff
@@ -934,7 +934,105 @@ impl MemPool {
...
+                current_size < size_limit
+            })
+            .collect();
+
+
+        let mut current_signed_tx: Vec<SignedTransaction> = pending_items.iter().map(|t| t.tx.clone()).collect();
+        let mut future_signed_tx: Vec<SignedTransaction> = future_pending_items.iter().map(|t| t.tx.clone()).collect();
+        for tx in &mut future_signed_tx {
+            current_signed_tx.push(tx.clone());
+        }
```
Comment (joojis) [inline core/src/miner/mem_pool.rs]: I think there is a better way to concatenating 2 vectors. but, I'm not sure at this moment.

You can look at https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.chain Or ask anybody else.
These days I don't write Rust code much. 😓

File: test/src/e2e.long/mempool.test.ts [hunk 6]
```diff
@@ -38,7 +38,10 @@ describe("Memory pool size test", function() {
             sending.push(nodeA.sendPayTx({ seq }));
         }
         await Promise.all(sending);
-        const pendingTransactions = await nodeA.sdk.rpc.chain.getPendingTransactions();
+        const pendingTransactions = await nodeA.sdk.rpc.sendRpcRequest(
+            "mempool_getPendingTransactions",
+            [null, null, true]
```
Comment (joojis) [inline test/src/e2e.long/mempool.test.ts]: Before this patch, the test would send `[null, null, false]`. I guess this was the cause of test failure.

File: core/src/miner/mem_pool.rs [hunk 7]
```diff
@@ -934,7 +934,106 @@ impl MemPool {
...
+            })
+            .collect();
+
+
+        let mut current_signed_tx: Vec<SignedTransaction> = pending_items.iter().map(|t| t.tx.clone()).collect();
+        let mut future_signed_tx: Vec<SignedTransaction> = future_pending_items.iter().map(|t| t.tx.clone()).collect();
+        current_signed_tx.append(&mut future_signed_tx);
+        //        for tx in &mut future_signed_tx {
+        //            current_signed_tx.push(tx.clone());
+        //        }
```
Comment (joojis) [inline core/src/miner/mem_pool.rs]: Would you remove the comments?

File: core/src/miner/mem_pool.rs [hunk 8]
```diff
@@ -934,7 +934,106 @@ impl MemPool {
...
+        let transactions: Vec<SignedTransaction> = current_signed_tx;
+        let last_timestamp = future_pending_items.into_iter().map(|t| t.inserted_timestamp).max();
```
Comment (joojis) [inline core/src/miner/mem_pool.rs]: I think this would not work since you changed it to use `append`.

`future pending items` is an empty list at this point.
